File: geekbang/algorithm/source-code/1/lc150 逆波兰表达式求值.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Long> s = new Stack<>();
        for (String token : tokens) {
            if (token.equals("+") || token.equals("-") || token.equals("*") || token.equals("/")) {
                long y = s.peek();
                s.pop();
                long x = s.peek();
                s.pop();
                long z = calc(x, y, token);
                s.push(z);
            } else {
                s.push(Long.parseLong(token)); // 字符串转化为整数
            }
        }
        return s.peek().intValue();
    }

    long calc(long x, long y, String op) {
        if (op.equals("+")) return x + y;
        if (op.equals("-")) return x - y;
        if (op.equals("*")) return x * y;
        if (op.equals("/")) return x / y;
        return 0;
    }
}

File: geekbang/algorithm/source-code/1/lc150 逆波兰表达式求值.py
class Solution:
    def calc(self, x, y, op):
        if op == "+":
            return x + y
        if op == "-":
            return x - y
        if op == "*":
            return x * y
        if op == "/":
            return int(x / y)
        return 0

    def evalRPN(self, tokens: List[str]) -> int:
        s = []
        for token in tokens:
            if token == "+" or token == "-" or token == "*" or token == "/":
                y = s[-1]
                s.pop()
                x = s[-1]
                s.pop()
                z = self.calc(x, y, token)
                s.append(z)
            else:
                s.append(int(token))
        return s[-1]